### DISCUSSION TITLE
To use function_exists() or not

### DISCUSSION THREAD

**Discussion by Aetles** [gregsullivan/_tw#59] [Q&A] [answered]
So I was trying to determine if the use of `function_exists()` was best practice or unnecessary. I'm talking about checking if a function exist before declaring a function, like this:
```php
if ( ! function_exists( '_tw_setup' ) ) :
	function _tw_setup() {
		...
	}
endif;
add_action( 'after_setup_theme', '_tw_setup' );
```
From what I've read this is mostly done as a "pluggable function", when a theme expect that it will be used as a parent for child themes, where the child theme can override the functions in the parent theme. A common comment was that pluggable functions generally should be avoided anyway, and it's better to use actions/filters. 

In the current `functions.php` there is one use of this, the other functions do not have this check, but in other files, like `inc/template-tags.php`, every functions has this check. 

Since _tw is not supposed to be a parent theme for child themes, but rather a starting theme, and since all the functions are prefixed, there seems to be little reason to use `functions_exists()`, right?

**Comment by gregsullivan** [comment]
Hi Adrian,

Sorry for the delay in responding! I was off last week, and I'm now getting caught up.

The one correction I'd make to what you wrote above is a somewhat subtle one:

> Since _tw is not supposed to be a parent theme for child themes, but rather a starting theme, and since all the functions are prefixed, there seems to be little reason to use `functions_exists()`, right?

_tw itself isn't meant to be used as a parent theme, but a theme created by the _tw theme generator could be used as a parent theme, and in that case, pluggable functions could be helpful. (The distinction here is that you're not meant to add _tw as-is in your theme directory planning to update it from the _tw repository as changes are made, with a child theme where you make your edits. You should generate a theme and edit it directly.)

This is a fairly narrow use case: Maybe someone is creating a parent theme for a multisite network, with child themes that override parts of the parent. There are other potential use cases, but generally along those lines.

Please let me know if that answers your question!

**Reply by Aetles** [reply]
Ah, yes I can see that use case. (And if it my use of _tw does not fall under that use case I could remove all the checks if I'd like to.)

But in that case, shouldn't all functions in `functions.php` have that check, or is it deliberate that only `_tw_setup()` is pluggable?

**Reply by gregsullivan** [reply, marked as answer]
Yes, you could absolutely remove those checks!

It is deliberate to the extent that not all functions need to be or should be pluggable. For example, the one that enqueues JS and CSS: Generally the idea with parent / child themes isn't that the parent theme provides no JS or CSS, but that the child theme supplements what the parent theme provides, so you probably wouldn't want that function to be pluggable. For the functions that add the Tailwind Typography classes, it should work to define the Typography constant in the child theme rather than overriding the functions that use the constant.

(There's certainly a world in which it's not deliberate because I overlooked a function that should be pluggable; I'll flag that as something to review in the future.)

To me, the canonical example of what we're discussing is Varia:

https://github.com/Automattic/themes/blob/trunk/varia/functions.php

Parts of _tw are based on Varia, and it is the parent theme to a number of Automattic's own themes. You can see that there are a mix of pluggable and non-pluggable functions, depending on the intent of the code within.

**Reply by Aetles** [reply]
Well, I was already impressed by _tw before and now I really feel that there's a great deal of thought behind it, thank you! 😊 I wasn't aware of Varia, nice to know!

I'm rebuilding a site currently based on Underscores and Foundation (which I think was a really good choice back then!) and _tw is exactly what I was looking for. And I'm trying to absorb every current best practice for a theme, hence the questions. 😄

If you don't mind me getting a bit off topic – I could open another discussion but since you mentioned Varia I took a look at something else I was thinking about: 

In Underscores they used `esc_html__()` on translated strings like `'menu-1' => esc_html__( 'Primary', '_s' ),` while _tw just have `'menu-1' => __( 'Primary', '_tw' ),`, so I was going to ask about it. I did see some old discussion [when they added it to Twenty Twenty](https://core.trac.wordpress.org/ticket/54127) where some thought it was ok to trust translations while others preferred to sanitize (they eventually added it). But now when I look at Varia I see that they don't escape and I guess that's why _tw doesn't either.

**Reply by gregsullivan** [reply]
You're very welcome—I appreciate that a lot!

The lack of escaping in those instances basically mirrors the sentiment in [this comment](https://core.trac.wordpress.org/ticket/30724#comment:13):

> Translations are inherently trusted. The `__()` family of functions are used thousands of times and they don't escape output. If we're not trusting translations then we have a big problem.

Additionally, since every generated theme gets its own text domain for translations, you should be in full control of your own translations. (The risk vector here is basically that someone puts HTML inside a translation that was just a word before—"Primary," say. But if you have your own text domain, you shouldn't need to worry about untrusted translations whatsoever.)

All of that said, over escaping isn't necessarily harmful, and I think WordPress has moved more in that direction to limit the amount of under escaping that is done by developers who might not fully understand the risks of echoing unescaped variables.

**Reply by Aetles** [reply]
Thanks again for a thorough answer, helping me in the right direction! 🙏

**Reply by gregsullivan** [reply]
You're very welcome!
